hub/api/tensor.py

```diff
@@ -1,35 +1,68 @@
 from hub.util.slice import merge_slices
 import numpy as np
 from typing import Union
+from hub.core.typing import StorageProvider
+from hub.core.storage.memory import MemoryProvider
+from hub.core.storage.local import LocalProvider
+from hub.core.chunk_engine.read import read_array, read_tensor_meta
+from hub.core.chunk_engine.write import write_array, write_tensor_meta
 
 
 class Tensor:
-    def __init__(self, uuid: str, tensor_slice: slice = slice(None)):
+    def __init__(
+        self,
+        key: str,
+        provider: StorageProvider,
+        tensor_slice: slice = slice(None),
+    ):
         """Initialize a new tensor.
 
         Note:
-            This operation does not create a new tensor in the backend,
+            This operation does not create a new tensor in the storage provider,
             and should normally only be performed by Hub internals.
 
         Args:
-            uuid (str): The internal identifier for this tensor.
-            tensor_slice (slice, optional): The slice object restricting the view of this tensor.
+            key (str): The internal identifier for this tensor.
+            provider (StorageProvider): The storage provider for the parent dataset.
+            tensor_slice (slice): The slice object restricting the view of this tensor.
         """
-        self.uuid = uuid
+        self.key = key
+        self.provider = provider
         self.slice = tensor_slice
-        self.shape = (0,)  # Dataset should pass down relevant metadata
+
+        self.load_meta()
+
+    def load_meta(self):
+        meta = read_tensor_meta(self.key, self.provider)
+        self.num_samples = meta["length"]
+        self.shape = meta["max_shape"]
 
     def __len__(self):
         """Return the length of the primary axis"""
-        return self.shape[0]
+        return self.num_samples
 
     def __getitem__(self, item: Union[int, slice]):
         if isinstance(item, int):
             item = slice(item, item + 1)
 
         if isinstance(item, slice):
             new_slice = merge_slices(self.slice, item)
-            return Tensor(self.uuid, new_slice)
+            return Tensor(self.key, self.provider, new_slice)
+
+    def __setitem__(self, item: Union[int, slice], value: np.ndarray):
```

Comment:
so is `__setitem__` the only way to append data? i thought we would allow something like:

```
ds = Dataset(...)
ds.append(np.array([1,2,3,4])
```

and have it automatically use `write_tensor` and `append_tensor`, raising errors if the `ds.mode` is incompatible for each operation.

Reply:
it seems like to do this functionality i would have to:

```
ds = Dataset(...)
ds["tensor"] = np.array([1,2,3])
ds["tensor"][1] = np.array([4,5,6])
```

Reply:
or if you don't like the idea of appending to a non-existent tensor, then we can can do something like this instead:

```
ds = Dataset(...)
ds.create_tensor("tensor")
ds["tensor"].append(np.array([1,2,3]))
ds["tensor"].append(np.array([4,5,6]))  # later
```

i personally like this approach the most

Reply:
@AbhinavTuli @imshashank @benchislett  what do you guys think?

Reply:
This is exactly what I meant. No, setitem is not planned to be used to append at all.